streams/src/main/java/org/apache/kafka/streams/state/internals/VersionedRecordIterator.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import java.util.List;
+import java.util.NoSuchElementException;
+import org.apache.kafka.streams.state.ValueIterator;
+import org.apache.kafka.streams.state.VersionedRecord;
+
+public class VersionedRecordIterator<V> implements ValueIterator<VersionedRecord<V>> {
+
+    protected final List<VersionedRecord<V>> records;
+    protected int currentIndex;
+
+    public VersionedRecordIterator(final List<VersionedRecord<V>> records) {
+        this.records = records;
+        this.currentIndex = 0;
+    }
+
+    @Override
+    public void close() {
+
+    }
+
+    @Override
+    public VersionedRecord<V> peek() {
+        return next();
```

Comment:
peek() should not advance the iterator, and thus cannot call next().

Can we re-use `AbstractIterator` and only implement `makeNext()` ?

Reply:
> peek() should not advance the iterator, and thus cannot call next().
> 
> Can we re-use `AbstractIterator` and only implement `makeNext()` ?

Then we do not need the `ValueIterator` interface?

Reply:
As discussed in person -- we still need `ValueIterator` because we need to implement `Closable`.